File: lang/java-to-kotlin/src/main/java/com/projectkorea/java/step01/Main.java
package com.projectkorea.java.step01;

// 퍼블릭 클래스가 있는 경우, 그 클래스의 이름과 파일 이름이 일치해야 한다.

public class Main {
    // public: 접근 제어자
    // 다른 클래스에서 호출할 수 있음.
    // main 메서드는 자바 애플리케이션의 시작점이므로 반드시 public이어야함

    // static: 키워드
    // main 메서드는 프로그램의 시작점으로, 자바 런타임에 의해 호출
    // 인스턴스를 생성하지 않고 클래스 자체에서 호출할 수 있음을 의미합니다.
    // 자바 런타임은 클래스의 객체를 생성하지 않고 main 메서드를 호출하기 때문에 static이어야 한다..

    //  void: 반환 타입
    //	이 메서드는 값을 반환하지 않음을 의미합니다.
    //	main 메서드는 애플리케이션의 진입점이므로 실행을 마친 후 별도의 값을 반환할 필요가 없다.

    // main: 메서드 이름
    // 이름은 main으로 고정
    // Java 진입점으로 인식되는 메서드, 자바 프로그램이 시작될 때 자동으로 호출되
    // 시그니처가 정확히 맞아야 Java 런타임 환경(JVM)이 프로그램을 시작할 수 있다.

    public static void main(String[] args){
        long num1 = 10L;
        final long num2 = 10L;
        Long num3 = 1_000L; // 레퍼런스 타입이기에 null도 가능

        // 1. String str2 = num2.toString(); // Cannot call method because 'num2' has primitive type long
        String str3 = num3.toString();

        // 2. primitive, Reference 타입을 섞어서 연산 시, 불필요한 프로세스가 진행

        // 원시 타입과 래퍼 타입을 섞어서 연산 (박싱 발생)
        Long sumNum = num1 + num2 + num3;

        // 결과를 원시 타입으로 사용 (언박싱 발생)
        long finalNum = sumNum;

        Person person = new Person("junha");
        System.out.println(person.getName());
    }
}

File: lang/java-to-kotlin/src/main/kotlin/com/projectkorea/kotlin/step01/Main.kt
package com.projectkorea.kotlin.step01

// 1.
// Kotlin에서는 파일 이름과 클래스 이름이 일치할 필요가 없습니다. 파일 내에 여러 클래스나 함수를 정의할 수 있습니다.

// 2.
// Kotlin에서는 원시 타입(primitive types)이 아니라 모든 것이 객체입니다.
// 기본 타입들은 Java의 원시 타입들과 대응되며, Kotlin 표준 라이브러리에서 제공하는 클래스들로 래핑되어 있습니다.

// 3. import com.projectkorea.step01.Person
// 같은 package 안에 있으면 import 할 필요 없음

fun main() {
    val num1 = 10L
    // 1.
    // val num1 : Long = 10L
    // type inference 지원

    // 2.
    // val == const

    val num2 = 10L
    var num3: Long? = 1_000L
    // 2.
    // 기본적으로 모든 변수는 null이 들어갈 수 없도록 설계
    // nullable 하다면 type?를 명시적으로 사용하자

    val num4 = num1 + num2   // + num3; type misMatch
    // 4.
    // 연산시 알아서 Long타입이 long으로 바뀌며 효율적으로 계산

    num3 = 2_000_000L
    num3 = null

    val person = Person("junha")
    // 5. new 키워드 생략
    // 6. 자바, 코틀린 문자열은 " 따옴표 사용

    println(person.name)
}


// 요약
// 1.타입 명시하지 않아도 추론 가능
// 2. Primitive, Reference Type 구분 안해도 됌
    // 모든 타입을 객체로 취급.
    // 코틀린이 원시 타입을 객체로 포장하는 과정(오토박싱과 언박싱)을 자동으로 처리
// 3. nullable 타입은 ?만 붙이고, 완전히 타입으로 구분된다.
// 4. 객체의 인스턴스화는 new를 제외한다.
// 5. tools/kotlin -> "show kotlin byteCode" ->  Decompile
    // Java 파일로 변환된 것을 볼 수 있음.